drgn_tools/slabinfo.py

```diff
@@ -0,0 +1,310 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helper to view slabinfo data
+"""
+import argparse
+from typing import Set
+
+import drgn
+from drgn import cast
+from drgn import sizeof
+from drgn.helpers.linux.cpumask import for_each_online_cpu
+from drgn.helpers.linux.list import list_for_each_entry
+from drgn.helpers.linux.percpu import per_cpu_ptr
+from drgn.helpers.linux.slab import for_each_slab_cache
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.table import print_table
+
+PAGE_MAPPING_ANON = 1
+output = []
+
+
+def collect_node_info(
+    prog: drgn.Program,
+    s: drgn.Object,
+    nrnodeid: int,
+    total_slabs: int,
+    free_objects: int,
+    total_inuse: int,
+) -> None:
+    """
+    collect per node slab data and
+    print the collective output
+    """
+    nr_slabs = 0
+    nr_total_objs = 0
+    nr_free = 0
+    nr_partial = 0
+    node_total_use = 0
+    full_slabs = 0
+    cpu_per_node = 0
+    ad = hex(s.value_())
+    ssize = int((int(prog["PAGE_SIZE"]) << int(s.oo.x >> 16)) / 1024)
+
+    for nodeid in range(nrnodeid):
+        kmem_cache_node = s.node[nodeid]
+        x = 0
+        nr_slabs = nr_slabs + kmem_cache_node.nr_slabs.counter.value_()
+        nr_total_objs = (
+            nr_total_objs + kmem_cache_node.total_objects.counter.value_()
+        )
+        nr_partial = nr_partial + kmem_cache_node.nr_partial.value_()
+        cpu_per_node = 0
+        try:
+            for page in list_for_each_entry(
+                "struct page",
+                kmem_cache_node.partial.address_of_(),
+                "slab_list",
+            ):
+                nrobj = page.objects.value_()
+                nrinuse = page.inuse.value_()
+                x = nrobj - nrinuse
+                node_total_use += nrinuse
+                nr_free += x
+        except LookupError:
+            for page in list_for_each_entry(
+                "struct page", kmem_cache_node.partial.address_of_(), "lru"
+            ):
+                nrobj = page.objects.value_()
+                nrinuse = page.inuse.value_()
+                x = nrobj - nrinuse
+                node_total_use += nrinuse
+                nr_free += x
```

Comment:
You're repeating the body of the loop for both conditions here. At some point, somebody is going to update this code and forget to update the other copy of it.

It would be better to have something like `for_each_partial_slab()` or something, which detects the proper way to iterate over the slabs, and returns the appropriate generator. That way you can avoid duplicating code.